node/directives/validateUserAccess.ts

```diff
@@ -0,0 +1,122 @@
+import { AuthenticationError, ForbiddenError } from '@vtex/api'
+import type { GraphQLField } from 'graphql'
+import { defaultFieldResolver } from 'graphql'
+import { SchemaDirectiveVisitor } from 'graphql-tools'
+
+import type { AuthAuditTokenMetrics } from '../metrics/auth'
+import sendAuthMetric, { AuthMetric } from '../metrics/auth'
+import {
+  validateAdminToken,
+  validateApiToken,
+  validateStoreToken,
+} from './helper'
+
+export class ValidateUserAccess extends SchemaDirectiveVisitor {
+  public visitFieldDefinition(field: GraphQLField<any, any>) {
+    const { resolve = defaultFieldResolver } = field
+
+    field.resolve = async (
+      root: any,
+      args: any,
+      context: Context,
+      info: any
+    ) => {
+      const {
+        vtex: { adminUserAuthToken, storeUserAuthToken, logger },
+      } = context
+
+      let hasTokens = true
+      let hasValidTokens = true
+
+      const { hasAdminToken, hasValidAdminToken } = await validateAdminToken(
+        context,
+        adminUserAuthToken as string
+      )
+
+      // add Admin token metrics
+      let tokenMetrics: AuthAuditTokenMetrics = {
+        hasAdminToken,
+        hasValidAdminToken,
+      }
+
+      if (!hasAdminToken || !hasValidAdminToken) {
+        const { hasApiToken, hasValidApiToken } = await validateApiToken(
+          context
+        )
+
+        // add API token metrics
+        tokenMetrics = {
+          ...tokenMetrics,
+          hasApiToken,
+          hasValidApiToken,
+        }
+
+        if (!hasApiToken || !hasValidApiToken) {
+          const { hasStoreToken, hasValidStoreToken } =
+            await validateStoreToken(context, storeUserAuthToken as string)
+
+          // add store token metrics
+          tokenMetrics = {
+            ...tokenMetrics,
+            hasStoreToken,
+            hasValidStoreToken,
+          }
+
+          if (!hasStoreToken || !hasValidStoreToken) {
+            hasTokens = hasAdminToken || hasApiToken || hasStoreToken
+            hasValidTokens =
+              hasValidAdminToken || hasValidApiToken || hasValidStoreToken
+          }
+        }
+      }
```

Comment:
Can we avoid these nested `ifs`? Maybe we can construct the metric before everything and for each check, if it token is valid, we emit metric and return, otherwise e just update the metrics locally and go to the next check.

This should reduce complexity for this function.

Reply:
Fixed